keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1085 @@
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+Upgrade: set `recursiveReadOnly` in `[]volumeMounts`.
+
+Downgrade: unset `recursiveReadOnly` in `[]volumeMounts`.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+CRI has to be updated.
```

Comment:
What happens if APIserver is newer that kubelet?  A pod might Say "rro=Enabled" and the pod is running, but RRO is not set.  How does a user know?  E.g. "the status.containerStatuses[].volumeMounts[].rro field will not be set"

Reply:
Updated

```
- If kube-apiserver is older than the version that implements this KEP,
  kube-apiserver will generate an "unknown field" error.

- If kubelet is older than kube-apiserver, kubelet may ignore `[]volumeMounts.recursiveReadOnly`,
  or maybe generate an "unknown field" error, depending on its implementation.
  The user may check the existence of the `[]containerStatuses.[]volumeMount.recursiveReadOnly`
  property to see if it was ignored.

- If the CRI or the OCI runtime is older than kubelet, kubelet will raise an error
  if `RecursiveReadOnly` mode is set to `Enabled`.
  `IfPossible` will be treated as `Disabled`.
  The user may check the value of the `[]containerStatuses.[]volumeMount.recursiveReadOnly`
  property to see how it was treated.
```